src/core/pile.sol

```diff
@@ -17,150 +17,203 @@ pragma solidity >=0.4.24;
 
     // --- Data ---
-    TokenLike public tkn;
-    DebtRegisterLike public debtRegister;
-
     // https://github.com/makerdao/dsr/blob/master/src/dsr.sol
-    struct Loan {
-        uint balance;
-        uint rate;
+    struct Rate {
+        uint   pie;         // Total debt of all loans with this rate
+        uint   chi;         // Accumulated rates
+        uint   speed;       // Accumulation per second
+        uint48 rho;         // Last time the rate was accumulated
     }
 
-    mapping (uint => Loan) public loans_;
+    mapping (uint => Rate) public rates;
 
-    function loans(uint loan) public view returns (uint debt, uint balance, uint rate)  {
-        uint debt = debtRegister.debtOf(loan, loans_[loan].rate);
-        return (debt, loans_[loan].balance, loans_[loan].rate);
-    }
+    // loan => pie
+    // pie = debt/chi
+    mapping (uint => uint) public pie;
+    // loan => rate
+    mapping (uint => uint) public group;
 
-    function Debt() public view returns (uint debt) {
-        return debtRegister.totalDebt();
-    }
+    uint public total;
 
-    uint public Balance;
-    address public lender;
-
-    constructor(address tkn_, address title_, address debtRegister_) TitleOwned(title_) public {
+    constructor() public {
         wards[msg.sender] = 1;
-        tkn = TokenLike(tkn_);
-        debtRegister = DebtRegisterLike(debtRegister_);
+        rates[0].chi      = ONE;
+        rates[0].speed    = ONE;
+    }
+
+    // --- Public Debt Methods  ---
+    // increase the loan's debt by wad
+    function incDebt(uint loan, uint wad) public auth note {
+        uint rate = group[loan];
+        require(now <= rates[rate].rho);
+        uint delta = toPie(rates[rate].chi, wad);
+        pie[loan] = add(pie[loan], delta);
+        rates[rate].pie = add(rates[rate].pie, delta);
+        total = add(total, wad);
+    }
+
+    // decrease the loan's debt by wad
+    function decDebt(uint loan, uint wad) public auth note {
+        uint rate = group[loan];
+        require(now <= rates[rate].rho);
+        uint delta = toPie(rates[rate].chi, wad);
+        pie[loan] = sub(pie[loan], delta);
+        rates[rate].pie = sub(rates[rate].pie, delta);
+        total = sub(total, wad);
+    }
+
+    function debt(uint loan) public view returns (uint) {
+        uint rate = group[loan];
+        uint chi = rates[rate].chi;
+        if (now >= rates[rate].rho) {
+            (chi,) = compound(rate);
+        }
+        return fromPie(chi, pie[loan]);
+    }
+
+    function rateDebt(uint rate) public view returns (uint) {
+        uint chi = rates[rate].chi;
+        uint pie = rates[rate].pie;
+        
+        if (now >= rates[rate].rho) {
+            (chi,) = compound(rate);
+        } 
+        return fromPie(chi, pie);
+    } 
+
+    // --- Interest Rate Group Implementation ---
+
+    // set rate group for a loan
+    function setRate(uint loan, uint rate) public auth {
```

Comment:
That is exactly one of the questions I added to the discussion ticket

Reply:
added the check, as I also think it makes it safer to to mess up the rate group

Reply:
https://github.com/centrifuge/tinlake/pull/167/files/d8ccb527ceade2494e49e7b457f1d3cce130eca7#diff-e46d09b7a05c83ae1a2a8d09b1b9a8b6R52 

In this line we set the 0 rate group to just mean no interest. But with this change we imply that the rategroup is not set at all if it's 0. I think that will lead to issues. Do we want to forbid setting it to 0? Then we should fix the above code (either by deleting it because it's not needed or by moving it to rate group 1?)

Reply:
Sorry, I was actually referring to the file function that I think might require changes.  Currently when we changing the speed of a rate group we just resetting chi to 1, which leads to issue when the group already has debt.

For this function (setRate) I think we dont even need to change anything.  We can keep the 0 rate group and the  `require(pie[loan]==0)` check should be enough. Why not being able to change the rate group of a loan if there is no debt on it
@manuelpolzhofer, @lucasvo 
What do you think?

Reply:
agree with Alina. Discussed one on one